fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1062,6 +1056,329 @@ Future<Void> managementClusterRemoveTenantFromGroup(Transaction tr,
+			                      self->ctx.dataClusterMetadata.get(),
+			                      Optional<ClusterConnectionString>(self->connectionString),
+			                      updatedEntry);
+		}
+
+		TraceEvent("MarkedDataClusterRestoring").detail("Name", self->ctx.clusterName.get());
+
+		return true;
+	}
+
+	ACTOR static Future<Void> markClusterAsReady(RestoreClusterImpl* self, Reference<typename DB::TransactionT> tr) {
+		if (self->ctx.dataClusterMetadata.get().entry.clusterState == DataClusterState::RESTORING) {
+			DataClusterEntry updatedEntry = self->ctx.dataClusterMetadata.get().entry;
+			updatedEntry.clusterState = DataClusterState::READY;
+
+			updateClusterMetadata(tr,
+			                      self->ctx.clusterName.get(),
+			                      self->ctx.dataClusterMetadata.get(),
+			                      Optional<ClusterConnectionString>(self->connectionString),
+			                      updatedEntry);
+		}
+
+		TraceEvent("MarkedDataClusterReady")
+		    .detail("Name", self->ctx.clusterName.get())
+		    .detail("Version", tr->getCommittedVersion());
+
+		return Void();
+	}
+
+	ACTOR static Future<Void> markManagementTenantAsError(Reference<typename DB::TransactionT> tr,
+	                                                      TenantName tenantName) {
+		state Optional<TenantMapEntry> tenantEntry = wait(tryGetTenantTransaction(tr, tenantName));
+
+		if (!tenantEntry.present()) {
+			return Void();
+		}
+
+		tenantEntry.get().tenantState = TenantState::ERROR;
+		ManagementClusterMetadata::tenantMetadata().tenantMap.set(tr, tenantName, tenantEntry.get());
+		return Void();
+	}
+
+	ACTOR static Future<bool> getTenantsFromDataCluster(RestoreClusterImpl* self, Reference<ITransaction> tr) {
+		TenantNameRef begin = self->clusterName;
+		TenantNameRef end = "\xff\xff"_sr;
+		state Future<KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>>> tenantsFuture =
+		    TenantMetadata::tenantMap().getRange(tr, begin, end, CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER);
+		state KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>> tenants = wait(tenantsFuture);
+
+		if (!tenants.results.empty()) {
+			for (auto t : tenants.results) {
+				self->dataClusterTenantMap.emplace(t.first, t.second);
+			}
+		}
+
+		return !tenants.more;
+	}
+
+	ACTOR static Future<bool> getTenantIdNameIndexFromDataCluster(RestoreClusterImpl* self,
+	                                                              Reference<ITransaction> tr) {
+		uint64_t begin = 0;
+		uint64_t end = std::numeric_limits<uint64_t>::max();
+		state Future<KeyBackedRangeResult<std::pair<int64_t, TenantName>>> tenantsFuture =
+		    TenantMetadata::tenantIdIndex().getRange(tr, begin, end, CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER);
+		state KeyBackedRangeResult<std::pair<int64_t, TenantName>> tenants = wait(tenantsFuture);
+
+		if (!tenants.results.empty()) {
+			for (auto t : tenants.results) {
+				self->dataClusterTenantIdIndex.emplace(t.first, t.second);
```

Comment:
It looks like the only place this index is used is to check whether the ID exists at all (the name is not used), so optionally it could be made into a set. Also, I think you could build this map/set in `getTenantsFromDataCluster` since we have all the information we need there, and that way you could avoid the range scan.